File: graphlib-3/go/lib/components.go
package graphlib

import (
	"sort"
)

// HasCycle detects if the graph contains a cycle
func HasCycle(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"has_cycle": false, "cycle": []string{}}
	}

	if len(g.Nodes) == 0 {
		return map[string]interface{}{"has_cycle": false, "cycle": []string{}}
	}

	if g.Directed {
		return hasCycleDirected(g)
	}
	return hasCycleUndirected(g)
}

func hasCycleDirected(g *Graph) map[string]interface{} {
	// Use DFS with coloring: 0 = white (unvisited), 1 = gray (in progress), 2 = black (done)
	color := make(map[string]int)
	parent := make(map[string]string)

	var cycle []string
	hasCycle := false

	var dfs func(node string) bool
	dfs = func(node string) bool {
		color[node] = 1 // gray

		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if color[neighbor] == 1 {
				// Back edge found - cycle detected
				hasCycle = true
				// Reconstruct cycle
				cycle = []string{neighbor}
				current := node
				for current != neighbor {
					cycle = append([]string{current}, cycle...)
					current = parent[current]
				}
				cycle = append(cycle, neighbor) // Complete the cycle
				return true
			}
			if color[neighbor] == 0 {
				parent[neighbor] = node
				if dfs(neighbor) {
					return true
				}
			}
		}

		color[node] = 2 // black
		return false
	}

	// Get sorted nodes for deterministic order
	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if color[node] == 0 {
			if dfs(node) {
				break
			}
		}
	}

	if !hasCycle {
		return map[string]interface{}{"has_cycle": false, "cycle": []string{}}
	}
	return map[string]interface{}{"has_cycle": true, "cycle": cycle}
}

func hasCycleUndirected(g *Graph) map[string]interface{} {
	visited := make(map[string]bool)
	parent := make(map[string]string)

	var cycle []string
	hasCycle := false

	var dfs func(node, par string) bool
	dfs = func(node, par string) bool {
		visited[node] = true

		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				parent[neighbor] = node
				if dfs(neighbor, node) {
					return true
				}
			} else if neighbor != par {
				// Back edge found - cycle detected
				hasCycle = true
				// Reconstruct cycle
				cycle = []string{neighbor}
				current := node
				for current != neighbor {
					cycle = append([]string{current}, cycle...)
					current = parent[current]
				}
				cycle = append(cycle, neighbor) // Complete the cycle
				return true
			}
		}
		return false
	}

	// Get sorted nodes for deterministic order
	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if !visited[node] {
			if dfs(node, "") {
				break
			}
		}
	}

	if !hasCycle {
		return map[string]interface{}{"has_cycle": false, "cycle": []string{}}
	}
	return map[string]interface{}{"has_cycle": true, "cycle": cycle}
}

// IsDAG checks if the graph is a Directed Acyclic Graph
func IsDAG(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"is_dag": false}
	}

	// Undirected graphs are not DAGs
	if !g.Directed {
		return map[string]interface{}{"is_dag": false}
	}

	// Empty directed graph is a DAG
	if len(g.Nodes) == 0 {
		return map[string]interface{}{"is_dag": true}
	}

	// A directed graph is a DAG if it has no cycles
	cycleResult := hasCycleDirected(g)
	return map[string]interface{}{"is_dag": !cycleResult["has_cycle"].(bool)}
}

// TopologicalSort returns a topological ordering of the graph
func TopologicalSort(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"success": false, "error": "graph_not_found"}
	}

	// Check if it's a DAG
	if !g.Directed {
		return map[string]interface{}{"success": false, "error": "not_a_dag"}
	}

	cycleResult := hasCycleDirected(g)
	if cycleResult["has_cycle"].(bool) {
		return map[string]interface{}{"success": false, "error": "not_a_dag"}
	}

	// Kahn's algorithm
	inDegree := make(map[string]int)
	for node := range g.Nodes {
		inDegree[node] = 0
	}
	for node := range g.Nodes {
		for neighbor := range g.Adj[node] {
			inDegree[neighbor]++
		}
	}

	// Queue of nodes with no incoming edges
	queue := []string{}
	for node := range g.Nodes {
		if inDegree[node] == 0 {
			queue = append(queue, node)
		}
	}
	sort.Strings(queue)

	order := []string{}
	for len(queue) > 0 {
		// Sort queue to ensure deterministic order
		sort.Strings(queue)
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)

		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	return map[string]interface{}{"success": true, "order": order}
}

// ConnectedComponents finds all connected components
func ConnectedComponents(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"count": 0, "components": [][]string{}}
	}

	if len(g.Nodes) == 0 {
		return map[string]interface{}{"count": 0, "components": [][]string{}}
	}

	visited := make(map[string]bool)
	components := [][]string{}

	// Get sorted nodes for deterministic order
	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	for _, node := range nodes {
		if !visited[node] {
			component := []string{}
			bfsComponent(g, node, visited, &component)
			sort.Strings(component)
			components = append(components, component)
		}
	}

	return map[string]interface{}{"count": len(components), "components": components}
}

func bfsComponent(g *Graph, start string, visited map[string]bool, component *[]string) {
	queue := []string{start}
	visited[start] = true

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		*component = append(*component, node)

		// For connectivity, we need to consider both directions for directed graphs (weak connectivity)
		neighbors := make(map[string]bool)
		for neighbor := range g.Adj[node] {
			neighbors[neighbor] = true
		}
		for neighbor := range g.InAdj[node] {
			neighbors[neighbor] = true
		}

		sortedNeighbors := make([]string, 0, len(neighbors))
		for neighbor := range neighbors {
			sortedNeighbors = append(sortedNeighbors, neighbor)
		}
		sort.Strings(sortedNeighbors)

		for _, neighbor := range sortedNeighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}
}

// StronglyConnectedComponents finds SCCs using Kosaraju's algorithm
func StronglyConnectedComponents(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"count": 0, "components": [][]string{}}
	}

	if len(g.Nodes) == 0 {
		return map[string]interface{}{"count": 0, "components": [][]string{}}
	}

	// For undirected graphs, SCC is the same as CC
	if !g.Directed {
		return ConnectedComponents(graphID)
	}

	// Kosaraju's algorithm
	// Step 1: Fill nodes in stack according to their finishing times
	visited := make(map[string]bool)
	stack := []string{}

	nodes := make([]string, 0, len(g.Nodes))
	for node := range g.Nodes {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	var fillOrder func(node string)
	fillOrder = func(node string) {
		visited[node] = true

		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				fillOrder(neighbor)
			}
		}
		stack = append(stack, node)
	}

	for _, node := range nodes {
		if !visited[node] {
			fillOrder(node)
		}
	}

	// Step 2: Create a reversed graph (use InAdj)
	// Step 3: Do DFS using the reversed graph in order of decreasing finish time
	visited = make(map[string]bool)
	components := [][]string{}

	var dfsReverse func(node string, component *[]string)
	dfsReverse = func(node string, component *[]string) {
		visited[node] = true
		*component = append(*component, node)

		neighbors := make([]string, 0, len(g.InAdj[node]))
		for neighbor := range g.InAdj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				dfsReverse(neighbor, component)
			}
		}
	}

	// Process nodes in reverse order of finishing time (pop from stack)
	for i := len(stack) - 1; i >= 0; i-- {
		node := stack[i]
		if !visited[node] {
			component := []string{}
			dfsReverse(node, &component)
			sort.Strings(component)
			components = append(components, component)
		}
	}

	return map[string]interface{}{"count": len(components), "components": components}
}

// IsConnected checks if the graph is connected
func IsConnected(graphID string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"is_connected": false}
	}

	// Empty graph is connected
	if len(g.Nodes) == 0 {
		return map[string]interface{}{"is_connected": true}
	}

	// Single node is connected
	if len(g.Nodes) == 1 {
		return map[string]interface{}{"is_connected": true}
	}

	// Check weak connectivity (same as number of connected components == 1)
	ccResult := ConnectedComponents(graphID)
	return map[string]interface{}{"is_connected": ccResult["count"].(int) == 1}
}

File: graphlib-3/ts/src/components.ts
import {
  HasCycleResult,
  IsDAGResult,
  TopologicalSortResult,
  ConnectedComponentsResult,
  IsConnectedResult,
} from './types';
import { getGraph } from './graph';

export function hasCycle(graphId: string): HasCycleResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { has_cycle: false, cycle: [] };
  }

  if (graph.directed) {
    return hasCycleDirected(graphId);
  } else {
    return hasCycleUndirected(graphId);
  }
}

function hasCycleDirected(graphId: string): HasCycleResult {
  const graph = getGraph(graphId)!;
  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map<string, number>();
  const parent = new Map<string, string>();

  for (const node of graph.nodes) {
    color.set(node, WHITE);
  }

  let cycleStart: string | null = null;
  let cycleEnd: string | null = null;

  function dfs(node: string): boolean {
    color.set(node, GRAY);

    const neighbors = graph!.adjacency.get(node);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (color.get(neighbor) === GRAY) {
          // Back edge found - cycle detected
          cycleStart = neighbor;
          cycleEnd = node;
          return true;
        }
        if (color.get(neighbor) === WHITE) {
          parent.set(neighbor, node);
          if (dfs(neighbor)) {
            return true;
          }
        }
      }
    }

    color.set(node, BLACK);
    return false;
  }

  for (const node of graph.nodes) {
    if (color.get(node) === WHITE) {
      if (dfs(node)) {
        // Reconstruct cycle
        const cycle: string[] = [cycleStart!];
        let current = cycleEnd!;
        while (current !== cycleStart) {
          cycle.push(current);
          current = parent.get(current)!;
        }
        cycle.push(cycleStart!);
        cycle.reverse();
        return { has_cycle: true, cycle };
      }
    }
  }

  return { has_cycle: false, cycle: [] };
}

function hasCycleUndirected(graphId: string): HasCycleResult {
  const graph = getGraph(graphId)!;
  const visited = new Set<string>();
  const parent = new Map<string, string>();

  let cycleStart: string | null = null;
  let cycleEnd: string | null = null;

  function dfs(node: string, par: string | null): boolean {
    visited.add(node);

    const neighbors = graph!.adjacency.get(node);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (!visited.has(neighbor)) {
          parent.set(neighbor, node);
          if (dfs(neighbor, node)) {
            return true;
          }
        } else if (neighbor !== par) {
          // Back edge found
          cycleStart = neighbor;
          cycleEnd = node;
          return true;
        }
      }
    }

    return false;
  }

  for (const node of graph.nodes) {
    if (!visited.has(node)) {
      if (dfs(node, null)) {
        // Reconstruct cycle
        const cycle: string[] = [cycleStart!];
        let current = cycleEnd!;
        while (current !== cycleStart) {
          cycle.push(current);
          current = parent.get(current)!;
        }
        cycle.push(cycleStart!);
        return { has_cycle: true, cycle };
      }
    }
  }

  return { has_cycle: false, cycle: [] };
}

export function isDAG(graphId: string): IsDAGResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { is_dag: false };
  }

  // DAG only applies to directed graphs
  if (!graph.directed) {
    return { is_dag: false };
  }

  // Empty directed graph is a DAG
  if (graph.nodes.size === 0) {
    return { is_dag: true };
  }

  const cycleResult = hasCycle(graphId);
  return { is_dag: !cycleResult.has_cycle };
}

export function topologicalSort(graphId: string): TopologicalSortResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { success: false, error: 'graph_not_found' };
  }

  if (!graph.directed) {
    return { success: false, error: 'not_a_dag' };
  }

  const cycleResult = hasCycle(graphId);
  if (cycleResult.has_cycle) {
    return { success: false, error: 'not_a_dag' };
  }

  // Kahn's algorithm
  const inDegree = new Map<string, number>();
  for (const node of graph.nodes) {
    inDegree.set(node, 0);
  }

  for (const [, neighbors] of graph.adjacency) {
    for (const neighbor of neighbors.keys()) {
      inDegree.set(neighbor, (inDegree.get(neighbor) || 0) + 1);
    }
  }

  // Use sorted order for deterministic output
  const queue: string[] = [];
  for (const node of Array.from(graph.nodes).sort()) {
    if (inDegree.get(node) === 0) {
      queue.push(node);
    }
  }

  const order: string[] = [];

  while (queue.length > 0) {
    queue.sort();
    const current = queue.shift()!;
    order.push(current);

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        const newDegree = inDegree.get(neighbor)! - 1;
        inDegree.set(neighbor, newDegree);
        if (newDegree === 0) {
          queue.push(neighbor);
        }
      }
    }
  }

  return { success: true, order };
}

export function connectedComponents(graphId: string): ConnectedComponentsResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { count: 0, components: [] };
  }

  const visited = new Set<string>();
  const components: string[][] = [];

  // For directed graphs, use weak connectivity (ignore direction)
  function getNeighbors(node: string): string[] {
    const neighbors = new Set<string>();
    const outNeighbors = graph!.adjacency.get(node);
    if (outNeighbors) {
      for (const n of outNeighbors.keys()) {
        neighbors.add(n);
      }
    }
    const inNeighbors = graph!.inEdges.get(node);
    if (inNeighbors) {
      for (const n of inNeighbors.keys()) {
        neighbors.add(n);
      }
    }
    return Array.from(neighbors);
  }

  for (const startNode of Array.from(graph.nodes).sort()) {
    if (!visited.has(startNode)) {
      const component: string[] = [];
      const queue: string[] = [startNode];
      visited.add(startNode);

      while (queue.length > 0) {
        const current = queue.shift()!;
        component.push(current);

        for (const neighbor of getNeighbors(current)) {
          if (!visited.has(neighbor)) {
            visited.add(neighbor);
            queue.push(neighbor);
          }
        }
      }

      components.push(component.sort());
    }
  }

  return { count: components.length, components };
}

export function stronglyConnectedComponents(
  graphId: string
): ConnectedComponentsResult {
  const graph = getGraph(graphId);
  if (!graph || graph.nodes.size === 0) {
    return { count: 0, components: [] };
  }

  // For undirected graphs, SCC equals CC
  if (!graph.directed) {
    return connectedComponents(graphId);
  }

  // Kosaraju's algorithm
  const visited = new Set<string>();
  const finishOrder: string[] = [];

  // First DFS to get finish order
  function dfs1(node: string): void {
    visited.add(node);
    const neighbors = graph!.adjacency.get(node);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (!visited.has(neighbor)) {
          dfs1(neighbor);
        }
      }
    }
    finishOrder.push(node);
  }

  for (const node of Array.from(graph.nodes).sort()) {
    if (!visited.has(node)) {
      dfs1(node);
    }
  }

  // Second DFS on transposed graph
  visited.clear();
  const components: string[][] = [];

  function dfs2(node: string, component: string[]): void {
    visited.add(node);
    component.push(node);
    // Use inEdges for transposed graph
    const neighbors = graph!.inEdges.get(node);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (!visited.has(neighbor)) {
          dfs2(neighbor, component);
        }
      }
    }
  }

  // Process in reverse finish order
  for (let i = finishOrder.length - 1; i >= 0; i--) {
    const node = finishOrder[i];
    if (!visited.has(node)) {
      const component: string[] = [];
      dfs2(node, component);
      components.push(component.sort());
    }
  }

  return { count: components.length, components };
}

export function isConnected(graphId: string): IsConnectedResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { is_connected: true };
  }

  if (graph.nodes.size <= 1) {
    return { is_connected: true };
  }

  const cc = connectedComponents(graphId);
  return { is_connected: cc.count === 1 };
}
